src/extensions/history-autocapture.ts

```diff
@@ -0,0 +1,129 @@
+import { PostHog } from '../posthog-core'
+import { window } from '../utils/globals'
+import { addEventListener } from '../utils'
+import { logger } from '../utils/logger'
+import { patch } from './replay/rrweb-plugins/patch'
+
+/**
+ * This class is used to capture pageview events when the user navigates using the history API (pushState, replaceState).
+ * It is also used to capture pageview events when the user navigates using the browser's back/forward buttons.
+ */
+export class HistoryAutocapture {
+    private _instance: PostHog
+    private _initialized = false
+    private _popstateListener: (() => void) | undefined
+
+    constructor(instance: PostHog) {
+        this._instance = instance
+        this.startIfEnabled()
+    }
+
+    /**
+     * Whether history events capture is enabled based on client configuration
+     */
+    public get isEnabled(): boolean {
+        return Boolean(this._instance.config.capture_history_events)
+    }
+
+    /**
+     * Start monitoring History API based on configuration
+     */
+    public startIfEnabled(): void {
+        if (this.isEnabled && !this._initialized) {
+            logger.info('History API monitoring enabled, starting...')
+            this.monitorHistoryChanges()
+        }
+    }
+
+    /**
+     * Stop monitoring History API events and clean up listeners
+     */
+    public stop(): void {
+        this._popstateListener?.()
+        this._popstateListener = undefined
+
+        this._initialized = false
+
+        logger.info('History API monitoring stopped')
+    }
+
+    /**
+     * Set up listeners for History API methods to capture pageview events for SPA navigation
+     */
+    public monitorHistoryChanges(): void {
+        if (!window || !window.history) {
+            return
+        }
+
+        // Capture PostHog instance reference for use in closure
+        const instance = this._instance
+        const isEnabled = () => this.isEnabled
+
+        // Only patch if not already patched
+        if (!(window.history.pushState as any)?.__posthog_wrapped__) {
+            patch(window.history, 'pushState', (originalPushState: any) => {
+                return function (this: History, ...args: any[]) {
+                    const result = originalPushState.apply(this, args)
+                    try {
+                        if (isEnabled()) {
+                            instance.capture('$pageview', { navigation_type: 'pushState' })
```

Comment:
Just thinking through some edge cases here.

What about something like e.g. metabase where every query run updates the state? Or sites that automatically update query params when you visit (I've seen this with google ad links, e.g. when I google shoes and click the first shopping link https://arterton.co.uk/products/acme-wingtip-balmoral-oxford-8091?variant=43255266050221&utm_medium=product_sync&utm_source=google&utm_content=sag_organic&utm_campaign=sag_organic&gad_source=1&gclid=EAIaIQobChMIxsXc4PvNjAMVlJJQBh0Dew1OEAQYASABEgJ18vD_BwE it did this (it added the country code as a query param)

Maybe this should only trigger when the actual pathname changes, rather than the query or hash? 

Or maybe those users just shouldn't use this 🤷

Reply:
That is a really good point. I think most use cases would be for only the pathname, but on that same link, if I click on different sizes, it changes the query for the variant, which some people may also want to track. 

We could add an option on the config like `{ captureHistoryChange: 'always' | 'pathname' | 'never'}` instead of just a boolean but I am afraid this could grow...

Thoughts @ivanagas ?

Reply:
I think we should only trigger a pageview on path change if possible and keep it as a boolean. 

This seems to be what people expect. For example, the Docusaurus plugin captured when the hash changed unintentionally and people complained so [we fixed it](https://github.com/PostHog/posthog-docusaurus/pull/22). Hash or state changes are probably handled better by autocapture or custom events.

Reply:
Cool! I've updated both PRs to follow that line and only capture pageviews when the path	 changes.